### ISSUE TITLE
ExternalInputIterator takes too much time

### ISSUE THREAD

**Issue by 674106399** [NVIDIA/DALI#2645]
Hi, I'm trying to use DALI to accelerate my training, but I find it's much slower than pure pytorch dataloader(about 2 times slower). I'm doing a pose estimation task, which requires 5 inputs during training, so I choose to use ExternalInputIterator  according to the example.

Here's my input iterator:

```python3
class ExternalInputIterator(object):
    def __init__(self, ds, batch_size):
        self.batch_size = batch_size
        self.ds = ds
        self.indices = list(range(len(self.ds)))
        shuffle(self.indices)

    def __iter__(self):
        self.i = 0
        self.n = len(self.ds)
        return self

    def __next__(self):
        batch = []
        target_coord, hand_type, gesture_label, img_arr = [], [], [], []   

        ## target_coord is a (b, 21, 2) float array 
        ## hand_type, gesture_label are both (b, ) float array
        ## img_arr is an array to save imgs

        start_time = time.time()
        for _ in range(self.batch_size):
            index = self.indices[self.i]
            img, target = self.ds[index]
            batch.append(img.numpy())
            target_coord.append(np.array(target[0]))
            hand_type.append(np.array(target[1]))
            gesture_label.append(np.array(target[2]))
            img_arr.append(np.array(target[3]))
            self.i = (self.i + 1) % self.n

        print(time.time() - start_time)   ## takes about 4s per batch(bs = 256)

        return (batch, target_coord, hand_type, gesture_label, img_arr)

class ExternalSourcePipeline(Pipeline):
    def __init__(self, batch_size, eii, num_threads, device_id):
        super(ExternalSourcePipeline, self).__init__(batch_size, num_threads, device_id, seed=12)
        self.source = ops.ExternalSource(source=eii, num_outputs=5)
        self.build()

    def define_graph(self):
        images,  target_coord, hand_type, gesture_label, img_arr = self.source()
        return images, target_coord, hand_type, gesture_label, img_arr

class DALIGenericIterator(DALIGenericIterator):

    def __init__(self,
                 pipelines,
                 output_map,
                 size=-1,
                 reader_name=None,
                 auto_reset=False,
                 fill_last_batch=None,
                 dynamic_shape=False,
                 last_batch_padded=False,
                 last_batch_policy=LastBatchPolicy.FILL):
        super().__init__(pipelines, output_map, size=size, auto_reset=auto_reset, dynamic_shape=dynamic_shape, last_batch_policy=last_batch_policy)

    def __len__(self):
        batch_count = self._size // (self._num_gpus * self.batch_size)
        last_batch = 1 if self._last_batch_policy == LastBatchPolicy.FILL else 0
        return batch_count + last_batch
```

Here's my definition:

```python3
def train_dataloader(self):
        if self.cfg.dali:
            eii_train = ExternalInputIterator(self.train_set, self.cfg.batch_size)
            pipe_train = ExternalSourcePipeline(batch_size=self.cfg.batch_size, eii=eii_train, num_threads=self.cfg.num_thread, device_id=self.cfg.gpus[0])

            return DALIGenericIterator(
                        pipe_train,
                        ['inputs', 'target_coord', 'hand_type', 'gesture_label', 'img_arr'],
                        size=len(self.train_set),
                        auto_reset=True,
                        last_batch_policy=LastBatchPolicy.FILL
                    )
        else:
            return DataLoader(self.train_set, shuffle=True, batch_size=self.cfg.batch_size, pin_memory=True, drop_last=not self.test, num_workers=self.cfg.num_thread, prefetch_factor=2)
```

It takes about 1.7s per batch(batch size = 256) with pytorch dataloader, but about 5s with DALI.
I find the for loop in ExternalInputIterator takes about 4s.
I'm quite new to this, but I feel that this is not the right way to use DALI, could you give me any suggestion? 
BTW, I'm using dali 0.28
Thanks a lot

**Comment by JanuszL** [comment]
Hi,
ExternalSource doesn't scale with the number of CPU threads. There is a pending PR addressing this https://github.com/NVIDIA/DALI/pull/2543.
Is it possible to use multiple FileReaders or NumpyReaders in your case? You can create multiple of them, and use the `files` argument to pass any list of files. If you maintain the order between the readers and pass the same seed (in case of enabling shuffling) your data should be loaded in sync (each of 5 inputs).

**Comment by mzient** [comment]
What is `ds` and, consequently, what are img and target? If these are torch tensors or arrays of torch tensors, you can use them directly - there's no need to go through numpy.
Also, you can do away with ExternalInputIterator - you don't really need an iterator, you can do the followin:

```Python
import nvidia.dali as dali
import nvidia.dali.fn as fn
import torch


def get_data(sample_info):
    image = torch.tensor([[1,2],[3,4]], dtype=torch.uint8) + sample_info.idx_in_epoch
    label = torch.tensor(sample_info.idx_in_epoch, dtype=torch.uint8)
    return image, label

batch_size = 4
pipe = dali.pipeline.Pipeline(batch_size, 1, 0)
with pipe:
    image, label = fn.external_source(get_data, batch=False, num_outputs=2)
    pipe.set_outputs(image, label)

pipe.build()
for _ in range(5):
    images, labels = pipe.run()
    for i in range(batch_size):
        print("label: ", labels.at(i))
        print("image:\n", images.at(i), "\n")
```

Output:
```
label:  0
image:
 [[1 2]
 [3 4]] 

label:  1
image:
 [[2 3]
 [4 5]] 

label:  2
image:
 [[3 4]
 [5 6]] 

label:  3
image:
 [[4 5]
 [6 7]] 

label:  4
image:
 [[5 6]
 [7 8]] 

label:  5
image:
 [[6 7]
 [8 9]] 

label:  6
image:
 [[ 7  8]
 [ 9 10]] 

label:  7
image:
 [[ 8  9]
 [10 11]] 

label:  8
image:
 [[ 9 10]
 [11 12]] 

label:  9
image:
 [[10 11]
 [12 13]] 

label:  10
image:
 [[11 12]
 [13 14]] 

label:  11
image:
 [[12 13]
 [14 15]] 

label:  12
image:
 [[13 14]
 [15 16]] 

label:  13
image:
 [[14 15]
 [16 17]] 

label:  14
image:
 [[15 16]
 [17 18]] 

label:  15
image:
 [[16 17]
 [18 19]] 

label:  16
image:
 [[17 18]
 [19 20]] 

label:  17
image:
 [[18 19]
 [20 21]] 

label:  18
image:
 [[19 20]
 [21 22]] 

label:  19
image:
 [[20 21]
 [22 23]] 
```

**Comment by JanuszL** [comment]
Hi @674106399,

Please check the most recent improvements in [the `external_source` operator](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/supported_ops.html#nvidia.dali.fn.external_source) - especially `parallel` and `prefetch_queue_depth`. If you use them you should see an improvement in the performance.
